File: julia/day02.jl
using Chain

score_part1 = Dict((
    ((:rock, :rock), 1 + 3),
    ((:rock, :paper), 2 + 6),
    ((:rock, :scissors), 3 + 0),
    ((:paper, :rock), 1 + 0),
    ((:paper, :paper), 2 + 3),
    ((:paper, :scissors), 3 + 6),
    ((:scissors, :rock), 1 + 6),
    ((:scissors, :paper), 2 + 0),
    ((:scissors, :scissors), 3 + 3)
))

score_part2 = Dict((
    ((:rock, :lose), 3 + 0),
    ((:rock, :draw), 1 + 3),
    ((:rock, :win), 2 + 6),
    ((:paper, :lose), 1 + 0),
    ((:paper, :draw), 2 + 3),
    ((:paper, :win), 3 + 6),
    ((:scissors, :lose), 2 + 0),
    ((:scissors, :draw), 3 + 3),
    ((:scissors, :win), 1 + 6)
))

function transform_round_part1(round::Vector)
    rps = Dict([("A", :rock), ("B", :paper), ("C", :scissors), ("X", :rock), ("Y", :paper), ("Z", :scissors)])
    map(x -> rps[x], round)
end

function transform_round_part2(round::Vector)
    rps = Dict([("A", :rock), ("B", :paper), ("C", :scissors), ("X", :lose), ("Y", :draw), ("Z", :win)])
    map(x -> rps[x], round)
end

function parse_input(filename::String)
    @chain begin
        open("data/$filename", "r") do io
            read(io, String)
        end
        strip(_, '\n')
        split(_, '\n')
        map(split, _)
    end
end

function part1(filename::String)
    @chain begin
        parse_input(filename)
        map(transform_round_part1, _)
        map(x -> tuple(x...), _)
        map(x -> score_part1[x], _)
        sum(_)
    end
end

function part2(filename::String)
    @chain begin
        parse_input(filename)
        map(transform_round_part2, _)
        map(x -> tuple(x...), _)
        map(x -> score_part2[x], _)
        sum(_)
    end
end

@assert(part1("day02_sample.txt") == 15)
@assert(part1("day02.txt") == 12586)

@assert(part2("day02_sample.txt") == 12)
@assert(part2("day02.txt") == 13193)

File: rust/src/day02.rs
use anyhow::{anyhow, Error, Result};
use nom::{
    character::complete::{alpha1, line_ending, space1},
    combinator::map,
    multi::separated_list1,
    sequence::separated_pair,
    IResult,
};

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum Toss {
    Rock,
    Paper,
    Scissors,
}

impl From<Toss> for u32 {
    fn from(value: Toss) -> Self {
        match value {
            Toss::Rock => 1,
            Toss::Paper => 2,
            Toss::Scissors => 3,
        }
    }
}

impl TryFrom<&str> for Toss {
    type Error = Error;

    fn try_from(input: &str) -> Result<Toss> {
        match input {
            "X" | "A" => Ok(Toss::Rock),
            "Y" | "B" => Ok(Toss::Paper),
            "Z" | "C" => Ok(Toss::Scissors),
            x => Err(anyhow!("invalid input character: {:?} for Toss", x)),
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum Outcome {
    Lose,
    Draw,
    Win,
}

impl From<Outcome> for u32 {
    fn from(value: Outcome) -> Self {
        match value {
            Outcome::Lose => 0,
            Outcome::Draw => 3,
            Outcome::Win => 6,
        }
    }
}

impl TryFrom<&str> for Outcome {
    type Error = Error;

    fn try_from(input: &str) -> Result<Outcome> {
        match input {
            "X" => Ok(Outcome::Lose),
            "Y" => Ok(Outcome::Draw),
            "Z" => Ok(Outcome::Win),
            x => Err(anyhow!("invalid input character: {:?} for Outcome", x)),
        }
    }
}

#[allow(unused)]
struct Game {
    oppo: Toss,
    me: Toss,
    outcome: Outcome,
}

impl From<Game> for u32 {
    fn from(game: Game) -> Self {
        let toss_score: u32 = game.me.into();
        let outcome_score: u32 = game.outcome.into();
        toss_score + outcome_score
    }
}

impl Game {
    fn from_tosses(oppo: Toss, me: Toss) -> Game {
        let outcome = match (oppo, me) {
            (Toss::Rock, Toss::Rock) => Outcome::Draw,
            (Toss::Rock, Toss::Paper) => Outcome::Win,
            (Toss::Rock, Toss::Scissors) => Outcome::Lose,
            (Toss::Paper, Toss::Rock) => Outcome::Lose,
            (Toss::Paper, Toss::Paper) => Outcome::Draw,
            (Toss::Paper, Toss::Scissors) => Outcome::Win,
            (Toss::Scissors, Toss::Rock) => Outcome::Win,
            (Toss::Scissors, Toss::Paper) => Outcome::Lose,
            (Toss::Scissors, Toss::Scissors) => Outcome::Draw,
        };

        Game { oppo, me, outcome }
    }

    fn from_outcome(oppo: Toss, outcome: Outcome) -> Game {
        let me = match (oppo, outcome) {
            (Toss::Rock, Outcome::Win) => Toss::Paper,
            (Toss::Rock, Outcome::Lose) => Toss::Scissors,
            (Toss::Rock, Outcome::Draw) => Toss::Rock,
            (Toss::Paper, Outcome::Win) => Toss::Scissors,
            (Toss::Paper, Outcome::Lose) => Toss::Rock,
            (Toss::Paper, Outcome::Draw) => Toss::Paper,
            (Toss::Scissors, Outcome::Win) => Toss::Rock,
            (Toss::Scissors, Outcome::Lose) => Toss::Paper,
            (Toss::Scissors, Outcome::Draw) => Toss::Scissors,
        };

        Game { oppo, me, outcome }
    }
}

fn parse_part1(input: &str) -> IResult<&str, Vec<Game>> {
    separated_list1(line_ending, parse_lines_part1)(input)
}

fn parse_lines_part1(input: &str) -> IResult<&str, Game> {
    map(
        separated_pair(alpha1, space1, alpha1),
        |(o, m): (&str, &str)| {
            let oppo = Toss::try_from(o).unwrap();
            let me = Toss::try_from(m).unwrap();
            Game::from_tosses(oppo, me)
        },
    )(input)
}

fn parse_part2(input: &str) -> IResult<&str, Vec<Game>> {
    separated_list1(line_ending, parse_lines_part2)(input)
}

fn parse_lines_part2(input: &str) -> IResult<&str, Game> {
    map(
        separated_pair(alpha1, space1, alpha1),
        |(oppo, outcome): (&str, &str)| {
            let oppo = Toss::try_from(oppo).unwrap();
            let outcome = Outcome::try_from(outcome).unwrap();
            Game::from_outcome(oppo, outcome)
        },
    )(input)
}

#[allow(unused)]
fn part1(input: &'static str) -> Result<u32> {
    let (_, input) = parse_part1(input)?;
    let score: u32 = input.into_iter().map(|x| u32::from(x)).sum();
    Ok(score)
}

#[allow(unused)]
fn part2(input: &'static str) -> Result<u32> {
    let (_, input) = parse_part2(input)?;
    let score: u32 = input.into_iter().map(|x| u32::from(x)).sum();
    Ok(score)
}

#[cfg(test)]
mod tests {
    use super::*;

    const SAMPLE_TEXT: &str =
        include_str!("/home/steve/Documents/advent-of-code-2022/data/day02_sample.txt");
    const INPUT_TEXT: &str =
        include_str!("/home/steve/Documents/advent-of-code-2022/data/day02.txt");

    #[test]
    fn test_part1() {
        assert_eq!(part1(SAMPLE_TEXT).unwrap(), 15);
        assert_eq!(part1(INPUT_TEXT).unwrap(), 12586);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(SAMPLE_TEXT).unwrap(), 12);
        assert_eq!(part2(INPUT_TEXT).unwrap(), 13193);
    }
}
